### DISCUSSION TITLE
Is it safe to make several module instance from single module?

### DISCUSSION THREAD

**Discussion by palchikov** [bytecodealliance/wasm-micro-runtime#3896] [Q&A] [answered]
Hi there! In my scenario I want to have many modules instantiated from the same AOT file. They should be completely independent from each other, have their own memories and execution flow.
Currently I'm doing `wasm_runtime_load()` for each copy. And each `wasm_runtime_load()` makes `mmap` to first 2GB of host memory. Typical size of text section of my modules is 5-7MB and I want to have several hundred of module instances.
I'm wondering if I can create module via `wasm_runtime_load()` once and then create hundred instances via `wasm_runtime_instantiate()`? Will they still be completely isolated from each other?

**Comment by lum1n0us** [comment, marked as answer]
Generally speaking, a WebAssembly runtime creates instances as the execution form for a module. Each instance has its own linear memory, which is essentially equivalent to a separate running state. So `wasm_runtime_instantiate()` only should be OK.
